[FSDP] Ignore buffer type casting in ignored modules

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106766

issue resolved: https://github.com/pytorch/pytorch/issues/97791

before this PR, mixed_precision applies to buffers from ignored modules. see ```test_state_dict_with_ignored_modules(mixed_precision=True)``` for reproduce

after, we avoid applying mixed_precision semantics to buffers from ignored modules
* step 1 initialization: state._ignored_buffer_names contains all the buffers from ignored modules
* step 2 lazy init at runtime: skip ignored buffers in ```_get_buffers_and_dtypes_for_computation```
* step 3 skip upcasting in state_dict hook: avoid upcasting for ignored buffers in ```_get_buffers_and_dtypes_for_computation```

weifengpy committed:
[FSDP] Ignore buffer type casting in ignored modules

issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -1022,6 +1056,13 @@ def test_state_dict_with_ignored_modules(
                 f"{prefixed_tensor_name}",
             )
+        # should not apply mixed_precision to ignored buffers
+        for buffer_name in {
+            **buffer_to_buffer_name,
+        }.values():
```
nit: Can this be simplified to:
```suggestion
        for buffer_name in buffer_to_buffer_name.values():
```

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -635,6 +639,37 @@ def _get_ignored_params(
+    ignored_modules: Set[torch.nn.Module],
+) -> Set[str]:
+    """
+    Returns the buffers of the modules in ``ignored_modules``
```
```suggestion
    Returns the cleaned buffer FQNs of the modules in ``ignored_modules``
```

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -635,6 +639,37 @@ def _get_ignored_params(
+        }
+    )
+
+    # Always include nested FSDP modules' ignored buffers
```
nit:
```suggestion
    # Always include nested FSDP modules' ignored buffer names
```

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -111,6 +111,7 @@ def __init__(self) -> None:
         # FSDP/fully_shard.
         self._ignored_modules: Set[nn.Module] = set()
         self._ignored_params: Set[nn.Parameter] = set()
+        self._ignored_buffer_names: Set[str] = set()
```
```suggestion
        # Buffer names are cleaned (without wrapper prefixes)
        self._ignored_buffer_names: Set[str] = set()
```

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1485,10 +1486,15 @@ def _get_buffers_and_dtypes_for_computation(
                 continue
             visited_buffers.add(buffer)
+            if (
+                hasattr(fsdp_state, "_ignored_buffer_names")
```
Do we need the `hasattr(fsdp_state, "_ignored_buffer_names")` check? This attribute should always be defined?

weifengpy replied:
you are right. we do not need hasattr check. removed the check

awgu commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -250,8 +250,14 @@ def _common_unshard_post_state_dict_hook(
-            buffers.append(state_dict[fqn])
+            # skip upcasting for ignored buffers
+            if (
+                hasattr(fsdp_state, "_ignored_buffer_names")
```
Same with this `hasattr` check

awgu reviewed (approved):
This looks good to me! Thanks for the fast PR. I left some minor nits. The main thing is that for `FullyShardedDataParallel` (which is an `nn.Module` wrapper), we care about "clean" vs. "unclean" names, so it is good to be clear which kind of names we are using.

"Clean" means that any `nn.Module` wrapper prefixes have been removed (e.g. `_fsdp_wrapped_module.`). "Unclean" means they are left there untouched.

Also, JFYI, FQN stands for fully-qualified name, and this usually refers to the name returned from `named_parameters()` / `named_buffers()` when called on a root module. Sometimes though, it can be ambiguous which module the name is prefixed from, in which case it is good to make that clear in code/comments.

weifengpy commented:
> 

thanks for the context. FQN explanation really helped. Before this, I only get to understand it after pdb prints.

weifengpy committed:
Update on "[FSDP] Ignore buffer type casting in ignored modules"


issue resolved: https://github.com/pytorch/pytorch/issues/97791

before this PR, mixed_precision applies to buffers from ignored modules. see ```test_state_dict_with_ignored_modules(mixed_precision=True)``` for reproduce

after, we avoid applying mixed_precision semantics to buffers from ignored modules
* step 1 initialization: state._ignored_buffer_names contains all the buffers from ignored modules
* step 2 lazy init at runtime: skip ignored buffers in ```_get_buffers_and_dtypes_for_computation```
* step 3 skip upcasting in state_dict hook: avoid upcasting for ignored buffers in ```_get_buffers_and_dtypes_for_computation```

[ghstack-poisoned]

weifengpy commented:
@pytorchbot merge
